### PR TITLE
[#173530790] Adult Check on startBonusEligibilityCheck

### PR DESCRIPTION
**Pull request by AleDore**
Added check for logged user's age in order to make bonus request unavailable for not adult users.

### CODE REVIEW

File: src/services/bonusService.ts [hunk 1]
```diff
@@ -72,6 +75,15 @@ export default class BonusService {
     | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
   > =>
     withCatchAsInternalError(async () => {
+      // check if current logged user is enabled to start a bonus request
+      if (
+        fromNullable(user.date_of_birth).exists(
+          _ => !isOlderThan(18)(new Date(_), new Date())
+        )
+      ) {
+        return ResponseErrorUnexpectedAuthProblem();
```
Comment (gunzip) [inline src/services/bonusService.ts:84]: this returns 500. didn't we say 401 ? (can you add a test for this case ?)

Comment (AleDore) [reply src/services/bonusService.ts]: replaced 500 with 401 and test added ( similar to previous but I've changed the kind of response expected )

File: src/utils/__tests__/date.test.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+const toDate = new Date("2020-01-01");
+const olderThanValue = 18;
+
+describe("Check if a birthdate is for an adult user", () => {
+  it("should return true if the user has more than 18 years old", () => {
+    const validOlderDate = new Date("2000-01-01");
+    expect(isOlderThan(olderThanValue)(validOlderDate, toDate)).toBeTruthy();
+  });
+
+  it("should return true if the user has exactly 18 years old", () => {
```
Comment (gunzip) [inline src/utils/__tests__/date.test.ts:12]:
```suggestion
  it("should return true if the user is exactly 18 years old", () => {
```

File: src/utils/__tests__/date.test.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { isOlderThan } from "../date";
+
+const toDate = new Date("2020-01-01");
+const olderThanValue = 18;
+
+describe("Check if a birthdate is for an adult user", () => {
+  it("should return true if the user has more than 18 years old", () => {
```
Comment (gunzip) [inline src/utils/__tests__/date.test.ts:7]:
```suggestion
  it("should return true if the user is over 18 years old", () => {
```

File: src/utils/responses.ts [hunk 4]
```diff
@@ -68,3 +70,23 @@ export const withValidatedOrValidationError = <T, U>(
         errorsToReadableMessages(response.value).join(" / ")
       )
     : f(response.value);
+
+/**
+ * Interface for a forbidden error response.
```
Comment (gunzip) [inline src/utils/responses.ts:75]:
```suggestion
 * Interface for unauthorized error response.
```

File: src/utils/responses.ts [hunk 5]
```diff
@@ -68,3 +70,23 @@ export const withValidatedOrValidationError = <T, U>(
...
+ */
+export interface IResponseErrorUnauthorized
+  extends IResponse<"IResponseErrorUnauthorized"> {
+  readonly detail: string;
+}
+/**
+ * Returns a forbidden error response with status code 403.
```
Comment (gunzip) [inline src/utils/responses.ts:82]:
```suggestion
 * Returns an unauthorized error response with status code 401.
```

File: src/services/bonusService.ts [hunk 6]
```diff
@@ -69,9 +74,22 @@ export default class BonusService {
     | IResponseSuccessAccepted
     | IResponseErrorForbiddenNotAuthorized
     | IResponseErrorGone
+    | IResponseErrorUnauthorized
     | IResponseSuccessRedirectToResource<InstanceId, InstanceId>
   > =>
     withCatchAsInternalError(async () => {
+      // check if current logged user is enabled to start a bonus request
```
Comment (gunzip) [inline src/services/bonusService.ts:81]:
```suggestion
      // check if the current logged in user can start a bonus request
```

File: src/services/__tests__/bonusService.test.ts [hunk 7]
```diff
@@ -208,6 +222,19 @@ describe("BonusService#startBonusEligibilityCheck", () => {
   });
 });
 
+it("should return an error if logged user is not adult", async () => {
```
Comment (gunzip) [inline src/services/__tests__/bonusService.test.ts:225]:
```suggestion
it("should return an error if the logged in user is not adult", async () => {
```

**Pull request merged**